src/solvers/dgsem_tree/dg_2d_parabolic.jl

```diff
@@ -0,0 +1,652 @@
+  @unpack surface_flux = surface_integral
+
+  # Note: cache.boundaries.u contains the unsigned normal component (using "orientation", not "direction")
+  # of the viscous flux, as computed in `prolong2boundaries` of `calc_divergence!``
+  @unpack u, neighbor_ids, neighbor_sides, node_coordinates, orientations = cache.boundaries
+
+  @threaded for boundary in first_boundary:last_boundary
+    # Get neighboring element
+    neighbor = neighbor_ids[boundary]
+
+    for i in eachnode(dg)
+      # Get viscous boundary fluxes
+      flux_ll, flux_rr = get_surface_node_vars(u, equations_parabolic, dg, i, boundary)
+      if neighbor_sides[boundary] == 1 # Element is on the left, boundary on the right
+        flux_inner = flux_ll
+      else # Element is on the right, boundary on the left
+        flux_inner = flux_rr
+      end
+
+      x = get_node_coords(node_coordinates, equations_parabolic, dg, i, boundary)
+
+      # TODO: add a field in `cache.boundaries` for gradient information.
+      # Here, we pass in `u_inner = nothing` since we overwrite cache.boundaries.u with gradient information.
+      # This currently works with Dirichlet/Neuman boundary conditions for LaplaceDiffusion2D and
+      # NoSlipWall/Adiabatic boundary conditions for CompressibleNavierStokesEquations2D as of 2022-6-27.
+      # It will not work with implementations which utilize `u_inner` to impose boundary conditions.
+      flux = boundary_condition(flux_inner, nothing, get_unsigned_normal_vector_2d(direction),
+                                x, t, Divergence(), equations_parabolic)
+
+      # Copy flux to left and right element storage
+      for v in eachvariable(equations_parabolic)
+        surface_flux_values[v, i, direction, neighbor] = flux[v]
+      end
+    end
+  end
+
+  return nothing
+end
+
+function calc_gradient!(u_grad, u, t,
+                        mesh::Union{TreeMesh{2}, P4estMesh{2}}, equations_parabolic,
+                        boundary_conditions_parabolic, dg::DG, cache, cache_parabolic)
+  # Reset du
+  @trixi_timeit timer() "reset ∂u/∂t" begin
+    reset_du!(u_grad[1], dg, cache)
+    reset_du!(u_grad[2], dg, cache)
+  end
+
+  # Calculate volume integral
+  @trixi_timeit timer() "volume integral" begin
+    @unpack derivative_dhat = dg.basis
+    @threaded for element in eachelement(dg, cache)
+
+      # Calculate volume terms in one element
+      for j in eachnode(dg), i in eachnode(dg)
+        u_node = get_node_vars(u, equations_parabolic, dg, i, j, element)
+
+        for ii in eachnode(dg)
+          multiply_add_to_node_vars!(u_grad[1], derivative_dhat[ii, i], u_node, equations_parabolic, dg, ii, j, element)
+        end
+
+        for jj in eachnode(dg)
+          multiply_add_to_node_vars!(u_grad[2], derivative_dhat[jj, j], u_node, equations_parabolic, dg, i, jj, element)
+        end
+      end
+    end
+  end
+
+  # Prolong solution to interfaces
+  @trixi_timeit timer() "prolong2interfaces" begin
```

Comment:
Does the `cache_parabolic` support `eachinterface(dg, cache)`? If yes, we could replace this function by the regular `prolong2interfaces!` function, AFAICT

Reply:
Yep, it works. Done in f5f48fc04e822340e8613f62c2558c280beaafd3

What should we pass in for `surface_integral`?

Reply:
Right now, it is a kind of dummy argument there, I think. Thus, passing `nothing` and adding a comment there should work?

Reply:
Sounds good. Updated the comments

Reply:
Thanks! On that note: Does anyone know, *why* we do pass the surface integral into `prolong2interfaces!`? I randomly checked `prolong2interfaces!` functions from unstructured, p4est and DGMulti, and neither seems to use them... can we just remove them altogether 🙀?

Reply:
We need an additional surface integral for the FDSBP stuff on the `TreeMesh2D` - a strong form surface integral instead of the weak form.

Reply:
We planned to extend this to Gauss methods with different variants (standard weak form or Jesse's entropy-stable version) in the infrastructure of the `TreeMesh` etc. but never implemented it. Thus, it is sometimes just a dummy argument and only really needed in a few places.

Reply:
Thanks for the clarification! Maybe we should document this better, because right now we pass this into a lot of functions without it being needed.

Reply:
Could you please give it a shot? I implemented it, so I am the worst person to document it

Reply:
Done in 831f878e.